---
title: "2021-04-10_workshop"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r include=FALSE}
# You can use this chunk as a reference for the required libraries. If you think all libraries are installed, feel free to skip it.
#memory.limit(18000)
#install.packages(c('here', "Seurat", "patchwork", "dplyr", "sctransform",
#                   "tidyverse","ggplot2"))

#if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager") 
#BiocManager::install((c('AnnotationHub', 'ensembldb', 'multtest', 
#                          'ReactomeGSA', 'topGO', 'Rgraphviz')))
```

```{r include=FALSE}
# Loads libraries necessary for the script.
library(here)
library("Seurat")
library("patchwork")
library("dplyr")
library("sctransform")
library("ReactomeGSA")
library("topGO")
library("Rgraphviz")
library("cowplot")
library("tidyverse")
library("ggplot2")
```

```{r packages_check}
# Displays your loaded packages.
(.packages())
```

```{r load_data}
# To use the command "here" and refer to the implied structure of the project, make sure to create a new PROJECT, and not just a script/notebook. In Rstudio, go to "File" - "Create new project". The project should be located in the same folder, where "datasets" folder is located. Which, in turn, must only contain folders "hcc" and "healthy". Otherwise, you will receive an error here.

here()

data_dir_1 <- here("datasets", "hcc")
sample1_data <- Read10X(data.dir = data_dir_1)
HCC_sample <- CreateSeuratObject(counts = sample1_data, project = "HCC")

data_dir_2 <- here("datasets", "healthy")
sample2_data <- Read10X(data.dir = data_dir_2)
healthy_sample <- CreateSeuratObject(counts = sample2_data, project = "Healthy")
```

```{r QC}
# Quality control step. Don't forget to adjust or play around with filtering parameters 
# below to see how it impacts your analysis. Reminder: nFeature_RNA = number of 
# genes per cell, nCount_RNA = total number of molecules molecules per cell.

# healthy_mt
healthy_sample[["percent.mt"]] <- 
  PercentageFeatureSet(healthy_sample, pattern = "^MT-")
VlnPlot(healthy_sample, features = 
          c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

# HCC_mt
HCC_sample[["percent.mt"]] <- 
  PercentageFeatureSet(HCC_sample, pattern = "^MT-")
VlnPlot(HCC_sample, features = 
          c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)


# Filter_out
healthy_sample <- subset(healthy_sample, subset = nFeature_RNA > 200 
                         & nFeature_RNA < 2500 & percent.mt < 30)
HCC_sample <- subset(HCC_sample, subset = nFeature_RNA > 200 
                         & nFeature_RNA < 2500 & percent.mt < 30)
```

```{r data_integration}
# This is an option #1 (resource-light). I would personally prefer option #2 
# (the code block below). In case of option #2, ignore this block. How does option #1 work? 
# Here it is described: https://rdrr.io/cran/Seurat/man/NormalizeData.html
datasets.list <- list(healthy_sample, HCC_sample)

datasets.list <- lapply(X = datasets.list, FUN = function(x) {
    x <- NormalizeData(x)
    x <- FindVariableFeatures(x, selection.method = "vst", nfeatures = 2000)
})

features <- SelectIntegrationFeatures(object.list = datasets.list)

datasets.anchors <- FindIntegrationAnchors(object.list = datasets.list, anchor.features = features)
datasets.integrated <- IntegrateData(anchorset = datasets.anchors)
```

```{r data_integration_option_2, echo=FALSE}
# How does it work and why it might be better for you? Here is the reference: 
# https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1874-1
# This step may take some time.
#datasets.list <- list(healthy_sample, HCC_sample)
#for (i in 1:length(datasets.list)) {
#    datasets.list[[i]] <- SCTransform(datasets.list[[i]], verbose = FALSE)
#}
#datasets.features <- SelectIntegrationFeatures(object.list = datasets.list, nfeatures = 3000)
#datasets.list <- PrepSCTIntegration(object.list = datasets.list, anchor.features = datasets.features, 
#    verbose = FALSE)
#datasets.anchors <- FindIntegrationAnchors(object.list = datasets.list, normalization.method = "SCT", 
#    anchor.features = datasets.features, verbose = FALSE)
#datasets.integrated <- IntegrateData(anchorset = datasets.anchors, normalization.method = "SCT", 
#    verbose = FALSE)
```

```{r dim_reduction}
# If you have used option 1 above, then execute "lognormalize" section.
# If option 2, then ignore "lognormalize", uncomment (delete "#") section two,
# execute it. 

#lognormalize
datasets.integrated <- ScaleData(datasets.integrated, verbose = FALSE)
datasets.integrated <- RunPCA(datasets.integrated, npcs = 30, verbose = FALSE)
datasets.integrated <- RunUMAP(datasets.integrated, reduction = "pca", dims = 1:30)
datasets.integrated <- FindNeighbors(datasets.integrated, reduction = "pca", dims = 1:30)
datasets.integrated <- FindClusters(datasets.integrated, resolution = 0.5)
#sctransform
#datasets.integrated <- RunPCA(datasets.integrated, verbose = FALSE)
#datasets.integrated <- RunUMAP(datasets.integrated, dims = 1:30)
#plots <- DimPlot(datasets.integrated, group.by = c("tech", "celltype"), combine = FALSE)
#plots <- lapply(X = plots, FUN = function(x) x + theme(legend.position = "top") + guides(color = guide_legend(nrow = 3, 
#    byrow = TRUE, override.aes = list(size = 3))))
#CombinePlots(plots)
```

```{r umap_visualization}
# Also check the difference between group.by and split.by - try to change it in the first plot.
# Check whether cells between two conditions overlap.

p1 <- DimPlot(datasets.integrated, reduction = "umap", group.by = "orig.ident")
p2 <- DimPlot(datasets.integrated, reduction = "umap", label = TRUE, repel = TRUE)
p1 + p2
```

```{r annotation_1}
# This first option assumes that you know the exact markers for your cells.
# For fun, google names of genes which are markers or T or B cells and write them 
# instead of "gene1" and "gene2" (which are just placeholder - so it will not work unless changed).
# To play with plots, uncomment next lines and see which option would be optimal for you.
# After you assigned the cell names to the clasters (will be done later), you can return to this tep
# To check how exactly different clusters, e.g., two clusters of T-cells. For example,
# are they gdT-cells, or MAIT T-cells, or naive T-cells? Find it out, check literature for markers!

DefaultAssay(datasets.integrated) <- "RNA"
features <- c("gene1", "gene2")
#FeaturePlot(datasets.integrated, features = features, min.cutoff = "q9", label = TRUE)
#RidgePlot(datasets.integrated, features = features, ncol = 2)
#VlnPlot(datasets.integrated, features = features, split.by = "orig.ident")
#DotPlot(datasets.integrated, features = features) + RotatedAxis()
#DoHeatmap(subset(datasets.integrated, downsample = 100), features = features, size = 3)

```

```{r annotation_2}
# The second option - generate a .csv file to analyze in Rstudio or, for example, Microsoft Excel.
# Here you will receive a data table where you can filter for a cluster and find differentially expressed genes.
# Then use public databases (e.g. The Protein Atlas) to find out which cells usually express these markers.
# Nota bene: not all of the cells in this particular dataset are present in the "tissue" datasets on 
# The Protein Atlas. Some of them are of blood origin (e.g., infiltrating immune cells in hepatocellular
# carcinoma. That's why, you have to check the "blood cells" section as well.

#library(AnnotationHub)
#library(ensembldb)
#library(multtest)
#library(metap)
#BiocManager::install('limma')
#library('limma')

# Connect to AnnotationHub
#ah <- AnnotationHub()

# Access the Ensembl database for organism
#ahDb <- query(ah, 
#              pattern = c("Homo sapiens", "EnsDb"), 
#              ignore.case = TRUE)

# Acquire the latest annotation files
#id <- ahDb %>%
#        mcols() %>%
#        rownames() %>%
#        tail(n = 1)

# Download the appropriate Ensembldb database
#edb <- ah[[id]]

# Extract gene-level information from database
#annotations <- genes(edb, 
#                     return.type = "data.frame")

# Select annotations of interest
#annotations <- annotations %>%
#        dplyr::select(gene_id, gene_name, seq_name, gene_biotype, description)

#DefaultAssay(datasets.integrated) <-"RNA"
#get_conserved <- function(cluster){
#        FindConservedMarkers(datasets.integrated,
#                             ident.1 = cluster,
#                             grouping.var = "orig.ident",
#                             only.pos = TRUE) %>%
#                rownames_to_column(var = "gene") %>%
#                cbind(cluster_id = cluster, .)
#}

#map and receive dataframe
#conserved_markers <- FindAllMarkers(datasets.integrated, min.cells.group = 0)
#gene_descriptions <- unique(annotations[, c("gene_name", "description")])

#ann_conserved_markers <- left_join(x = conserved_markers,
#                                   y = gene_descriptions,
#                                   by = c("gene" = "gene_name"))
#write.csv(ann_conserved_markers,"ann_conserved_markers.csv")
```

```{r name_clusters}
# This is one of the options how you could put names to your clusters.
# Copy-paste it for each cluster (0, 1, 2, 3, etc.) and assign real names
# instead of "my_cool_cells" per cluster.
# Bonus task: find out/google how else it is possible to assign cell names
# (e.g., using metadata).
datasets.integrated <- RenameIdents(datasets.integrated, '0' = 'my_cool_cells')
```


```{r how_many_cells_per_cluster_do_we_have}
# To calculate how many cells per cluster do we have.
# Bonus task: make a table with columns for healthy and hcc samples.

cell.num <- table(Idents(datasets.integrated))
view(cell.num)
```

```{r reactome_pathways}
# Gene set enrichment analysis using the Reactome database.
# Reference: https://bioconductor.org/packages/release/bioc/vignettes/ReactomeGSA/inst/doc/analysing-scRNAseq.html
# First script is provided for example - you don't have to subset your cells.
# Instead, you can just put your whole Seurat object into the "analyze_sc_clusters" function.
# It will take more time, but results should be more interesting. 

my_awesome_cells <- subset(datasets.integrated, idents = c("my_cool_cells", "1"))
gsva_result <- analyse_sc_clusters(my_awesome_cells, verbose = TRUE)
pathway_expression <- pathways(gsva_result)

colnames(pathway_expression) <- gsub("\\.Seurat", "", colnames(pathway_expression))

pathway_expression[1:3,]

max_difference <- do.call(rbind, apply(pathway_expression, 1, function(row) {
    values <- as.numeric(row[2:length(row)])
    return(data.frame(name = row[1], min = min(values), max = max(values)))
}))
max_difference$diff <- max_difference$max - max_difference$min
# sort based on the difference
max_difference <- max_difference[order(max_difference$diff, decreasing = T), ]

head(max_difference)
plot_gsva_pathway(gsva_result, pathway_id = rownames(max_difference)[1])

plot_gsva_heatmap(gsva_result, max_pathways = 15, margins = c(6,20))

# limit to selected B cell related pathways
relevant_pathways <- c("R-HSA-983170", "R-HSA-388841", "R-HSA-2132295", "R-HSA-983705", "R-HSA-5690714")
plot_gsva_heatmap(gsva_result, 
                  pathway_ids = relevant_pathways, # limit to these pathways
                  margins = c(6,30), # adapt the figure margins in heatmap.2
                  dendrogram = "col", # only plot column dendrogram
                  scale = "row", # scale for each pathway
                  key = FALSE, # don't display the color key
                  lwid=c(0.1,4)) # remove the white space on the left

```

```{r go_analysis}
# Gene ontology analysis.
# Reference: https://ucdavis-bioinformatics-training.github.io/2017_2018-single-cell-RNA-sequencing-Workshop-UCD_UCB_UCSF/day3/scRNA_Workshop-PART6.html
# Bonus task: find out/google how to plot the results nicely.

BiocManager::install("org.Hs.eg.db")
library("org.Hs.eg.db")
cluster0 <- subset(datasets.integrated, idents = "0")
expr <- cluster0@assays$RNA@data
# Select genes that are expressed > 0 in at least 75% of cells (somewhat arbitrary definition)
n.gt.0 <- apply(expr, 1, function(x)length(which(x > 0)))
expressed.genes <- rownames(expr)[which(n.gt.0/ncol(expr) >= 0.75)]
all.genes <- rownames(expr)

# define geneList as 1 if gene is in expressed.genes, 0 otherwise
geneList <- ifelse(all.genes %in% expressed.genes, 1, 0)
names(geneList) <- all.genes

# Create topGOdata object
    GOdata <- new("topGOdata",
        ontology = "BP", # use biological process ontology
        allGenes = geneList,
        geneSelectionFun = function(x)(x == 1),
              annot = annFUN.org, mapping = "org.Hs.eg.db", ID = "symbol")

resultFisher <- runTest(GOdata, algorithm = "elim", statistic = "fisher")
allres <- GenTable(GOdata, Fisher = resultFisher, topNodes = 20, numChar = 60)

showSigOfNodes(GOdata, score(resultFisher), firstSigNode=5, useInfo ='all')
```

```
For reference on Seurat, use the official manuals: https://satijalab.org/seurat/articles/get_started.html
Datasets modified for the workshop are taken from the following publications:
HCC: Zhang Q, He Y, Luo N, Patel SJ, Han Y, Gao R, Modak M, Carotta S, Haslinger C, Kind D, Peet GW, Zhong G, Lu S, Zhu W, Mao Y, Xiao M, Bergmann M, Hu X, Kerkar SP, Vogt AB, Pflanz S, Liu K, Peng J, Ren X, Zhang Z. Landscape and Dynamics of Single Immune Cells in Hepatocellular Carcinoma. Cell. 2019 Oct 31;179(4):829-845.e20. doi: 10.1016/j.cell.2019.10.003. PMID: 31675496.
Healthy: MacParland SA, Liu JC, Ma XZ, Innes BT, Bartczak AM, Gage BK, Manuel J, Khuu N, Echeverri J, Linares I, Gupta R, Cheng ML, Liu LY, Camat D, Chung SW, Seliga RK, Shao Z, Lee E, Ogawa S, Ogawa M, Wilson MD, Fish JE, Selzner M, Ghanekar A, Grant D, Greig P, Sapisochin G, Selzner N, Winegarden N, Adeyi O, Keller G, Bader GD, McGilvray ID. Single cell RNA sequencing of human liver reveals distinct intrahepatic macrophage populations. Nat Commun. 2018 Oct 22;9(1):4383. doi: 10.1038/s41467-018-06318-7. PMID: 30348985; PMCID: PMC6197289.
```
